Setup husky

keento0809 opened the pull request:
### Feature Description

<!--- Describe your changes in detail -->

* Setup basic features of husky that are implemented every time git commit is executed (Typecheck for TypeScript and showing fix options for errors in our code for ES-Lint)

### Related Issue

N/A

<!--- Put issue number if it's related. ex) #8 -->
<!--- Delete it if not needed -->

### Something you want to mention

* I haven't set up husky for git push command yet. So if it's necessary to run husky when we implement git push, I'll create another branch addressing to add more settings for git push in husky.

<!--- Make a note about some problems or questions -->
<!--- Delete it if not needed -->

### Checklist before requesting a review

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

keento0809 committed:
Added the basic settings of husky

keento0809 committed:
Reset settings for husky for some errors

keento0809 committed:
test commit

keento0809 committed:
test commit again

keento0809 committed:
delete unused file

keento0809 committed:
fixing error regarding jsx

keento0809 committed:
Created lint-staged.config.js to solve error regarding lint-staged

keento0809 committed:
Testing lint-staged.config.js

keento0809 committed:
Testing lint-staged.config.js again

keento0809 committed:
Resolved conflicts

keento0809 committed:
Reset node_modules

keento0809 committed:
Added pre-push setting in husky

keento0809 committed:
Deleted pre-push settings in husky

keento0809 committed:
Merge branch 'develop' of https://github.com/keento0809/map-t3 into setup-husky

keento0809 committed:
Deleted unnecessary page file and comments

Lada496 commented on .eslintrc.cjs:
```diff
@@ -15,9 +15,7 @@ const config = {
-  parserOptions: {
-    project: path.join(__dirname, "tsconfig.json"),
-  },
+  parserOptions: {},
```
I'm not so familiar with TypeScript eslint stuff so please correct me if I'm doing wrong but I think we can keep the parserOptions.project to make the most of TypeScript power (https://typescript-eslint.io/linting/typed-linting) and just put ignorePatters for those we don't need to check with that plugin. (https://eslint.org/docs/latest/use/configure/ignore)
```suggestion
 project: path.join(__dirname, "tsconfig.json"),
      },
    },
  ],
  ignorePatterns: ["*.config.js"],
```

keento0809 replied:
@Lada496 I didn't exactly remember why I deleted parserOptions (probably to solve an error happened before), but yeah you're right! I'll add parserOptions as you suggested.

Lada496 replied:
@keento0809 Hope you can still see this comment. This is the original suggestion that I wanted to update.
instead of empty object, which I remember the inside of override, and just put ignorePatterns outside of override. 
That should be it.
I think the original one was set by t3 so I assume they are correct (maybe lol)

Lada496 commented:
@keento0809 thank you for setting every thing! I never tried setting up lint-staged by my self and I think you did amazing. I just left some suggestion so please take a look. Thanks!

keento0809 committed:
Update .eslintrc.cjs

Co-authored-by: Yuko Murayama <67321065+Lada496@users.noreply.github.com>

Lada496 commented on .eslintrc.cjs:
```diff
@@ -15,9 +15,11 @@ const config = {
-  parserOptions: {
-    project: path.join(__dirname, "tsconfig.json"),
-  },
+ project: path.join(__dirname, "tsconfig.json"),
```
I'm sorry but I meant the project property should be in parserOptions.
```suggestion
 parserOptions: {
        project: path.join(__dirname, "tsconfig.json"),
      },
    },
  ],
  ignorePatterns: ["*.config.js"],
```

Lada496 commented on .eslintrc.cjs:
```diff
@@ -15,9 +15,11 @@ const config = {
-  parserOptions: {
-    project: path.join(__dirname, "tsconfig.json"),
-  },
+ project: path.join(__dirname, "tsconfig.json"),
+      },
+    },
+  ],
+  ignorePatterns: ["*.config.js"],
```
I'm sorry but I meant the project property should be inside of parserOptions
```suggestion
parserOptions: {
        project: path.join(__dirname, "tsconfig.json"),
      },
    },
  ],
  ignorePatterns: ["*.config.js"],
```

keento0809 replied:
@Lada496 Oops! It's my bad. I just misunderstood that your previous suggestion contained project property inside parseOptions object

keento0809 committed:
Add project key-value inside parseOptions object

Co-authored-by: Yuko Murayama <67321065+Lada496@users.noreply.github.com>

Lada496 commented on .eslintrc.cjs:
```diff
@@ -15,9 +15,12 @@ const config = {
-  parserOptions: {
-    project: path.join(__dirname, "tsconfig.json"),
-  },
+parserOptions: {
+        project: path.join(__dirname, "tsconfig.json"),
+      },
+    },
+  ],
+  ignorePatterns: ["*.config.js"],
```
```suggestion
parserOptions: {
    project: path.join(__dirname, "tsconfig.json"),
  },
```

Lada496 replied:
Seems like we're screwed up with adding each part improperly.
Please make sure our changes apply this config settings

```
// eslint-disable-next-line @typescript-eslint/no-var-requires
const path = require("path");

/** @type {import("eslint").Linter.Config} */
const config = {
  overrides: [
    {
      extends: [
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
      ],
      files: ["*.ts", "*.tsx"],
      parserOptions: {
        project: path.join(__dirname, "tsconfig.json"),
      },
    },
  ],
  ignorePatterns: ["*.config.js"],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    project: path.join(__dirname, "tsconfig.json"),
  },
  plugins: ["@typescript-eslint"],
  extends: ["next/core-web-vitals", "plugin:@typescript-eslint/recommended"],
  rules: {
    "@typescript-eslint/consistent-type-imports": [
      "warn",
      {
        prefer: "type-imports",
        fixStyle: "inline-type-imports",
      },
    ],
    "@typescript-eslint/no-unused-vars": ["warn", { argsIgnorePattern: "^_" }],
  },
};

module.exports = config;
```

Lada496 commented on .eslintrc.cjs:
```diff
@@ -15,9 +15,12 @@ const config = {
     },
   ],
   parser: "@typescript-eslint/parser",
```
```suggestion
  ignorePatterns: ["*.config.js"],
  parser: "@typescript-eslint/parser",
```

Lada496 reviewed (commented):
Seems like we're screwed up and added each part improperly.
Please make sure config is the same as below before making new change

```
// eslint-disable-next-line @typescript-eslint/no-var-requires
const path = require("path");

/** @type {import("eslint").Linter.Config} */
const config = {
  overrides: [
    {
      extends: [
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
      ],
      files: ["*.ts", "*.tsx"],
      parserOptions: {
        project: path.join(__dirname, "tsconfig.json"),
      },
    },
  ],
  ignorePatterns: ["*.config.js"],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    project: path.join(__dirname, "tsconfig.json"),
  },
  plugins: ["@typescript-eslint"],
  extends: ["next/core-web-vitals", "plugin:@typescript-eslint/recommended"],
  rules: {
    "@typescript-eslint/consistent-type-imports": [
      "warn",
      {
        prefer: "type-imports",
        fixStyle: "inline-type-imports",
      },
    ],
    "@typescript-eslint/no-unused-vars": ["warn", { argsIgnorePattern: "^_" }],
  },
};

module.exports = config;

```

keento0809 commented:
> Seems like we're screwed up and added each part improperly. Please make sure config is the same as below before making new change
> 
> ```
> // eslint-disable-next-line @typescript-eslint/no-var-requires
> const path = require("path");
> 
> /** @type {import("eslint").Linter.Config} */
> const config = {
>   overrides: [
>     {
>       extends: [
>         "plugin:@typescript-eslint/recommended-requiring-type-checking",
>       ],
>       files: ["*.ts", "*.tsx"],
>       parserOptions: {
>         project: path.join(__dirname, "tsconfig.json"),
>       },
>     },
>   ],
>   ignorePatterns: ["*.config.js"],
>   parser: "@typescript-eslint/parser",
>   parserOptions: {
>     project: path.join(__dirname, "tsconfig.json"),
>   },
>   plugins: ["@typescript-eslint"],
>   extends: ["next/core-web-vitals", "plugin:@typescript-eslint/recommended"],
>   rules: {
>     "@typescript-eslint/consistent-type-imports": [
>       "warn",
>       {
>         prefer: "type-imports",
>         fixStyle: "inline-type-imports",
>       },
>     ],
>     "@typescript-eslint/no-unused-vars": ["warn", { argsIgnorePattern: "^_" }],
>   },
> };
> 
> module.exports = config;
> ```

@Lada496 I was a little confused about the setting of parserOptions and then didn't notice that I've added them inside overrides. I have to look into the differences between setting parserOptions inside overrides and setting them on the root level of config

Lada496 commented:
> 
@keento0809 
If you can find the original my suggestion, you might find what I wanted to do. That was just getting back project from the empty object and putting ignorePatterns right after override array.  (https://github.com/van-squad/map-t3/pull/4/files/0b8e325494910ef69e41c74f87616a7f9cfb249b)
I stupidly remove `parserOptions: {}` and put `project: path.join(__dirname, "tsconfig.json"),`, which broke everything.
Apologies for this confusion.
Hope it makes sense to you :)

keento0809 committed:
test ignorePatterns

keento0809 committed:
test ignorePatterns again

keento0809 committed:
Deleted unused codes

keento0809 committed:
Merge branch 'setup-husky' of https://github.com/keento0809/map-t3 into setup-husky

keento0809 committed:
Solved the error when pushing local branch

keento0809 commented:
> ignorePatterns: ["*.config.js"],

@Lada496 Sorry for my slow understanding of your suggestion, but now I got your proposal and pushed changes again.

Configuration setting of eslint doesn't seem to be a super big deal, but I found out that we can set .eslintignore file for ignoring eslint linting if it's necessary.

[https://eslint.org/docs/latest/use/configure/ignore#the-eslintignore-file](https://eslint.org/docs/latest/use/configure/ignore#the-eslintignore-file)

devjoylee reviewed (approved):
@keento0809  @Lada496  thank you for your hard work on the Husky setup!! 
Can't wait to try out the Husky environment on local and see how it will enhance our work 😊
